lms/djangoapps/courseware/courses.py

```diff
@@ -405,17 +409,50 @@ def get_course_date_blocks(course, user):
     if certs_api.get_active_web_certificate(course):
         block_classes.insert(0, CertificateAvailableDate)
 
-    blocks = (cls(course, user) for cls in block_classes)
+    blocks = [cls(course, user) for cls in block_classes]
+    blocks.extend(get_course_assignment_due_dates(
+        course, user, request, num_return=num_assignments, include_past_dates=include_past_dates))
 
-    def block_key_fn(block):
-        """
-        If the block's date is None, return the maximum datetime in order
-        to force it to the end of the list of displayed blocks.
-        """
-        if block.date is None:
-            return datetime.max.replace(tzinfo=pytz.UTC)
-        return block.date
-    return sorted((b for b in blocks if b.is_enabled), key=block_key_fn)
+    return sorted((b for b in blocks if b.date and (b.is_enabled or include_past_dates)), key=date_block_key_fn)
+
+
+def date_block_key_fn(block):
+    """
+    If the block's date is None, return the maximum datetime in order
+    to force it to the end of the list of displayed blocks.
+    """
+    return block.date or datetime.max.replace(tzinfo=pytz.UTC)
+
+
+def get_course_assignment_due_dates(course, user, request, num_return=None, include_past_dates=False):
+    """
+    Returns a list of assignment (at the subsection/sequential level) due date
+    blocks for the given course. Will return num_return results or all results
+    if num_return is None in date increasing order.
+    """
+    store = modulestore()
+    all_course_dates = get_dates_for_course(course.id, user)
+    date_blocks = []
+    for (block_key, date_type), date in all_course_dates.items():
+        if date_type == 'due' and block_key.block_type == 'sequential':
+            item = store.get_item(block_key)
```

Comment:
This is loading blocks one at a time from the modulestore. However, I think performance-wise it would be better if we could use the course-blocks api, which gets dates from edx-when, and also has info about whether the block has a due date and is a sequential. (Or, if course-blocks can just load a subset of blocks in the course, then you could query edx-when first, and then load the sequentials from the course-blocks api).

Reply:
When I was looking into that before, I remember it seemed like that way was going to be less efficient. I can't remember the exact details anymore, but I remember having issues trying to get the starting block usage key and then even after getting that and filtering for sequentials only, it still pulled in the whole course and then after the fact just got rid of anything that wasn't a sequential rather than only requesting sequentials. I also believe it didn't have all of the fields I wanted to request on, but I could be misremembering that part.

That being said, I'm very unclear of what black magic might be happening under the hood to make this more efficient in course-blocks than only querying once we know we want this particular block. It might help to talk through this more if you feel strongly about switching this. The logic here is largely based on what I saw in https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_api/api.py#L186

Reply:
The issue is that loading a single block out of the modulestore is expensive too. The course-blocks api has a bunch of things already precomputed, essentially, so as long as the fields that you need are already computed, it should be faster to make one call to the transformers.

@ormsbee, when we talked about the architecture of the block-transformer-like precursor thing, it was important for things like the transformers to be applied in consistent orders for caching. Is that still the case? For in-code uses, is the right thing to use `course_api/blocks/api.py`? Or is there another entry point that would be better?

Reply:
Okay folks, I'm sorry if I've delayed this PR with my concerns. A couple of things:

1. I think it's fine to ignore the potential perf hit here for now and see how it does after rollout.
2. Given where things are now, I think the cheapest way to get descriptors might be `store.get_items(course_key, qualifiers={'category': 'sequential'})`. But I'd honestly stick with the get_item() for now until it's a measured problem. Course blocks is more efficient for accessing thin slices of data cross a lot of nodes, but has a lot of up-front costs relative to getting a handful of sequences (because all that collected data for XBlock fields is mashed together).
3. In the longer run, if the only reason we need to touch the modulestore at all is to ask for the `graded` attribute, then I think that information should be put into edx-when during publish and come back in the `get_dates_for_course` call.

Reply:
@cpennington Are you good with this PR under Dave's thoughts above? To leave the code as is and monitor performance and if we see a hit, to update it to either use the `.get_items`, course_blocks api, or even update edx-when to hold the fields necessary for this?